src/phase/PhaseState.ts

```diff
@@ -119,6 +129,109 @@ export class PhaseState extends Struct({
     );
   }
 
+  applyRangedAttackAction(
+    action: Action,
+    actionSignature: Signature,
+    attackingPiece: Piece,
+    targetPiece: Piece,
+    attackingPieceWitness: PiecesMerkleWitness,
+    targetPieceWitness: PiecesMerkleWitness,
+    assertedAttackDistance: UInt32,
+    attackRoll: EncrytpedAttackRoll,
+    serverSecretKey: PrivateKey
+  ): PhaseState {
+    const piecesRoot = attackingPieceWitness.calculateRoot(
+      attackingPiece.hash()
+    );
+    piecesRoot.assertEquals(
+      targetPieceWitness.calculateRoot(targetPiece.hash()),
+      'Piece witnesses do not match each other'
+    );
+    piecesRoot.assertEquals(
+      this.currentPiecesState,
+      'Piece witnesses do not match the current game state'
+    );
+    this.playerPublicKey.assertEquals(attackingPiece.playerPublicKey);
+    attackingPiece.position
+      .verifyDistance(targetPiece.position, assertedAttackDistance)
+      .assertTrue('Asserted attack distance is not correct');
+    assertedAttackDistance.assertLessThanOrEqual(
+      attackingPiece.condition.rangedAttackRange,
+      'Asserted attack distance is out of range for attacking piece'
+    );
+    const v = actionSignature.verify(
+      this.playerPublicKey,
+      action.signatureArguments()
+    );
+    v.assertTrue('The action signature does not match the provided action');
+
+    action.nonce.assertGreaterThan(
+      this.actionsNonce,
+      'The action nonce is LTE the most recent action nonce'
+    );
+    action.actionType.assertEquals(
+      Field(1),
+      'Action type must be 1 (ranged attack)'
+    ); // action is a "ranged attack" action
+    action.actionParams.assertEquals(
+      targetPiece.hash(),
+      'The action target is different than the proved target piece'
+    );
+
+    const decrytpedRolls = attackRoll.decryptRoll(serverSecretKey);
+    // roll for hit
+    const hit = Circuit.if(
+      decrytpedRolls.hit.greaterThanOrEqual(
+        attackingPiece.condition.hitRoll.value
+      ),
+      Bool(true),
+      Bool(false)
+    );
+
+    // roll for wound
+    const wound = Circuit.if(
+      decrytpedRolls.wound.greaterThanOrEqual(
+        attackingPiece.condition.woundRoll.value
+      ),
+      Bool(true),
+      Bool(false)
+    );
+
+    // roll for save
+    const notSave = Circuit.if(
+      decrytpedRolls.save.greaterThanOrEqual(
+        attackingPiece.condition.saveRoll.value
+      ),
+      Bool(false),
+      Bool(true)
+    );
```

Comment:
Totally okay if we're leaving out armor piercing for now, but this will need to reference the `saveRoll` of the target piece, not that of the attacker.

Example:
```
saveRollThreshold = targetPiece.saveRoll + attackingPiece.armorPiercing
saveRollActual = rollDice()
saveRollPassed = saveRollActual >= saveRollThreshold
if (!saveRollPassed) applyDamage()
```
Just pseudocode and obviously not compatible with snarkyjs but hopefully that explains the general behavior

Reply:
Yes good catch!